cmd/provider/main.go

```diff
@@ -0,0 +1,172 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane-contrib/terrajet/pkg/cmdconfig"
+	"github.com/crossplane-contrib/terrajet/pkg/log"
+	"github.com/crossplane-contrib/terrajet/pkg/process"
+	"github.com/crossplane-contrib/terrajet/pkg/version"
+)
+
+const (
+	// command-line flags for nats-monitoring executable
+	flagDebug      = "debug"
+	flagTimeout    = "timeout"
+	flagPort       = "port"
+	flagPluginPath = "plugin-path"
+	// default values for command-line arguments
+	valTimeout    = 30 * time.Second
+	valPort       = 9000
+	valPluginPath = "/usr/local/bin/provider"
+	// config file path env. variable
+	appConfigFileEnvVarName = "APP_CONFIG_PATH"
+	// provider plugin binary path
+)
+
+// ProviderConfig maintains the configurations for the
+// provider runner
+type ProviderConfig struct {
+	// Debug enables debug level logging
+	Debug bool `mapstructure:"debug" yaml:"debug"`
+	// Timeout for reading the address info from provider's stdout
+	Timeout time.Duration `mapstructure:"timeout" yaml:"timeout"`
+	// Port the provider will listen on
+	Port int `mapstructure:"port" yaml:"port"`
+	// PluginPath path of the provider executable
+	PluginPath string `mapstructure:"plugin-path" yaml:"plugin-path"`
+}
+
+func (c ProviderConfig) String() (string, error) {
+	buff, err := yaml.Marshal(c)
+	if err != nil {
+		return "", err
+	}
+	return string(buff), nil
+}
+
+var logger logging.Logger
+
+func main() {
+	conf := &ProviderConfig{}
+	// initialize a cobra command for nats-monitoring
+	cmd := &cobra.Command{
+		Use:   "provider-wrapper",
+		Short: "Wrapper for provider",
+		Long:  "Wraps and manages provider plugin and socat processes",
+		PersistentPreRun: func(cmd *cobra.Command, _ []string) {
+			if _, err := cmdconfig.InitializeConfig(cmd, appConfigFileEnvVarName); err != nil {
+				panic(err)
+			}
+		},
+		Run: func(_ *cobra.Command, _ []string) {
+			startProvider(conf)
+		},
+	}
+
+	cmd.Flags().BoolVarP(&conf.Debug, flagDebug, "d", false, "Run with debug logging")
+	cmd.Flags().IntVarP(&conf.Port, flagPort, "p", valPort, "TCP port the provider will listen")
+	cmd.Flags().StringVar(&conf.PluginPath, flagPluginPath, valPluginPath,
+		"Path to the provider's plugin binary")
+	cmd.Flags().DurationVarP(&conf.Timeout, flagTimeout, "t", valTimeout,
+		"Timeout for reading provider's address")
+
+	if err := cmd.Execute(); err != nil {
+		panic(err)
+	}
+}
```

Comment:
We typically use https://github.com/alecthomas/kong (or its predecessor kingpin) for all Crossplane CLI parsing. I'd prefer not to introduce cobra unless there is a good reason to do so.

Reply:
Done. I have removed the provider wrapper implementation, which is for running a Terraform provider plugin as a shared gRPC server. I will also add some more details to the PR description about our current strategy for wrapping the Terraform CLI.